docs/architecture/adr-017-subspace-fee-grant.md

```diff
@@ -0,0 +1,261 @@
+
+### DeductFeeDecorator
+
+Currently, `x/auth` provides a `DeductFeeDecorator` based on `x/feegrant` to execute the action deducting the fees of a transaction from the signer/feepayer. We will build a new subspace-specified `DeductFeeDecorator` to replace the current one.
+
+The new subspace-specified `DeductFeeDecorator` will work as follows:
+1. check if all the messages in the transaction are messages related to the same subspace;
+2. if the messages in the transaction are all realted to the same subspace and a fee grant exists for that subspace, run the `x/subspaces` `DeductFeeDecorator`; otherwise run the `x/auth` `DeductFeeDecorator`,
+
+### Types
+
+#### Allowance
+
+The `x/feegrant` module already provides an interface (`FeeAllowanceI`) to represent a generic allowance, along with a set of useful implementations like `BasicAllowance`, `AllowedMsgAllowance` and `PeriodicAllowance`. Since we are not going to introduce new kinds of allowances, nor we are going to edit how an allowance is represented, we will reuse these types.
+
+#### User Grant
+
+The `x/feegrant` module provides a `Grant` object used to store the `granter`, `grantee` and what kind of `allowance` is granted to a specific user. Since this is enough information for us as well, we will use the same object to store user-related fee grants.
+
+#### Group Grant
+
+In order to represent a fee grant granted to a user group, we will implement a `GroupGrant` type. This will contain the `granter`, `group_id` and what kind of `allowance` is granted to the group:
+
+```protobuf
+message GroupGrant {
+    // granter is the address of the user granting an allowance of their funds.
+  string granter = 1 [(cosmos_proto.scalar) = "cosmos.AddressString"];
+
+  // the id of the group being granted an allowance of another user's funds.
+  uint32 group_id = 2;
+
+  // allowance can be any of basic, periodic, allowed fee allowance.
+  google.protobuf.Any allowance = 3 [(cosmos_proto.accepts_interface) = "FeeAllowanceI"];
+}
+```
+
+### Store
+
+#### Group grant
+
+In order to simplify granters to manage the group allowance, the granted group allowance will be stored in the keys having the structure like:
+```
+SubspaceGroupAllowancePrefix | SubspaceID | GranterAddress | GroupID | -> Protobuf(GroupGrant)
+```
+
+This structure allows granters to easily manage the group allowance inside a subspace by iterating over all allowances for the granters, which will be the most used query. In the other hand, grantees must know who the granter is when using the application, they can find their grant with O(N) time complexity, N is the number of granted groups by the granter.
```

Comment:
```suggestion
This structure allows granters to easily manage the group allowance inside a subspace by iterating over all allowances for the granters, which will be the most used query. On the other hand, it allows to perform the following queries: 
1. finding a specific grant given the granter address.   
   This can be done in O(N), where N is the number of granted groups by the granter.
2. finding all the grans given to a user based on the groups they are part of.
   This can be done in O(N * M), where N is the number of granters, and M is the number of grants each granter has given.

Considering that: 
- only users with permissions will be able to create grants in each subspace
- there will be a small number of grants each granter will create

then the time complexity of searching if a user has been granted a fee grant should not become too large.
```